### PR TITLE
Merging mk/fe_experiments into develop

### LINKED ISSUE #1470: Preparing mk/develop/fe_experiments to merge into develop
### Describe the task. It can be a feature, documentation, etc.

Matthias' branch mk/develop/fe_experiments needs to be merged into develop, however the code needs some clean up first, e.g. moving the plotting of latent tokens to a util file etc...

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by kctezcan**
## Description

This branch (mk/develop/fe_experiments) has been used for running all experiments for improving forecast skill. At this current stage it includes mostly i) changes in the default config, ii) addition of the layer norm, iii) ploting of weights/gradient norms

## Issue Number

Closes #1470

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/trainer.py [hunk 1]
```diff
@@ -718,15 +753,143 @@ def validate(self, mini_epoch):
...
+        model_state_dict = self.model.state_dict()
+        params = {
+            k: v
+            for k, v in params.items()
+            if k in model_state_dict and v.shape == model_state_dict[k].shape
+        }
+
+        is_model_sharded = self.cf.with_ddp and self.cf.with_fsdp
+        if is_model_sharded:
+            params = self.model.rename_old_state_dict(params=params)  # For backward compatibility
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Can we remove this.

File: config/default_config.yml [hunk 2]
```diff
@@ -10,7 +10,7 @@ embed_dropout_rate: 0.1
 target_cell_local_prediction: True
 
 ae_local_dim_embed: 1024
-ae_local_num_blocks: 2
+ae_local_num_blocks: 0
```
Comment (clessig) [inline config/default_config.yml]: Can we keep the default_config functionally identical and only add the new options but turned off

File: src/weathergen/utils/plot_grad_norms.py [hunk 3]
```diff
@@ -0,0 +1,525 @@
+import json
```
Comment (clessig) [inline src/weathergen/utils/plot_grad_norms.py]: Why does it appear here? We did merge the grad_norm PR onto develop before?

File: src/weathergen/train/trainer.py (hunk 1 above)
Comment (kctezcan) [reply src/weathergen/train/trainer.py]: sure

File: scripts/model_weight_progression.py
Comment (MatKbauer) [inline scripts/model_weight_progression.py:1]: Not sure whether we need this in develop. I'd leave it out.

File: config/default_config.yml (hunk 2 above)
Comment (MatKbauer) [reply config/default_config.yml]: @kctezcan just suggested to have a separate `dev_forecast_config.yml`. This sounds clean and I recall you also mentioned this, @clessig.

File: src/weathergen/utils/plot_grad_norms.py
Comment (MatKbauer) [inline src/weathergen/utils/plot_grad_norms.py:1]: Should also be removed.

File: config/default_config.yml [hunk 4]
```diff
@@ -42,18 +43,20 @@ pred_mlp_adaln: True
...
+forecast_policy: "fixed"
+forecast_freeze_model: False
 forecast_att_dense_rate: 1.0
-fe_num_blocks: 0
+fe_num_blocks: 8
 fe_num_heads: 16
 fe_dropout_rate: 0.1
 fe_with_qk_lnorm: True
+fe_layer_norm_after_blocks: []  # Index starts at 0. Thus, [3] adds a LayerNorm after the fourth layer
 impute_latent_noise_std: 0.0  # 1e-4
```
Comment (MatKbauer) [inline config/default_config.yml]: Let's eventually rename `Impute_latent_noise_std` to `fe_impute_latent_noise_std`. I had this on my list for some while.

File: config/default_config.yml [hunk 5]
```diff
@@ -42,18 +43,20 @@ pred_mlp_adaln: True
...
 
-healpix_level: 5
+healpix_level: 4
```
Comment (MatKbauer) [inline config/default_config.yml]: I think `healpix_level: 5` is what we should have in the forecast config. It changed back and forth recently, but 5 turns out to be better.

File: config/default_config.yml [hunk 6]
```diff
@@ -137,8 +140,8 @@ log_grad_norms: False
 
 start_date: 197901010000
 end_date: 202012310000
```
Comment (MatKbauer) [inline config/default_config.yml]: Change to `end_date: 202212310000`; I thought I had modified this but maybe forgot to push.

**Review by MatKbauer — COMMENTED** [review]
Added some comments to be modified until we can merge. Thanks @kctezcan!

**Comment by clessig** [comment]
Yes, let's have a second config for the forecasting. Thanks for the reminder.

File: src/weathergen/train/trainer.py (hunk 1 above)
Comment (kctezcan) [reply src/weathergen/train/trainer.py]: I removed all instances of this parameter renaming as well as the renaming function

File: packages/evaluate/src/weathergen/evaluate/io_reader.py [hunk 7]
```diff
@@ -7,40 +7,470 @@
...
+
 class WeatherGenReader(Reader):
     def __init__(self, eval_cfg: dict, run_id: str, private_paths: dict | None = None):
         """Data reader class for WeatherGenerator model outputs stored in Zarr format."""
 
         super().__init__(eval_cfg, run_id, private_paths)
 
         # TODO: remove backwards compatibility to "epoch" in Feb. 2026
-        self.mini_epoch = getattr(eval_cfg, "mini_epoch", getattr(eval_cfg, "epoch", -1))
+        self.mini_epoch = getattr(eval_cfg, "mini_epoch", eval_cfg["epoch"])
```
Comment (tjhunter) [inline packages/evaluate/src/weathergen/evaluate/io_reader.py]: `.get` please

File: src/weathergen/train/trainer.py [hunk 8]
```diff
@@ -496,6 +586,10 @@ def train(self, mini_epoch):
 
         self.optimizer.zero_grad()
 
+        # Unweighted loss, real weighted loss, std for losses that need it
+        self.loss_unweighted_hist, self.loss_model_hist, self.stdev_unweighted_hist = [], [], []
+        self.last_grad_norm = 0.0
```
Comment (tjhunter) [inline src/weathergen/train/trainer.py]: is this even used?

**Review by clessig — APPROVED** [review]
Thanks!
